library/general/src/lib/ui/multi_messages_dialog.rb

```diff
@@ -0,0 +1,247 @@
+    # @return [Symbol] User input or :timeout symbol
+    def user_input
+      if timed?
+        Yast::UI::TimeoutUserInput(TIMEOUT_STEP)
+      else
+        super
+      end
+    end
+
+    # Determines whether the dialog is timed or not
+    #
+    # @return [true,false] True if the dialog is timed; false otherwise.
+    def timed?
+      timeout.is_a?(Integer)
+    end
+
+  private
+
+    # Dialog options
+    #
+    # Overrides UI::Dialog#dialog_options
+    #
+    # @return [Yast::Term] Dialog options.
+    #
+    # @see UI::Dialog#dialog_options
+    def dialog_options
+      Opt(:decorated)
+    end
+
+    # Refresh message and buttons
+    #
+    # Update the message and enables/disables 'Next' and 'Back' buttons
+    # accordingly.
+    #
+    # @see #current_message
+    # @see #first_message?
+    # @see #last_message?
+    def refresh
+      Yast::UI.ReplaceWidget(Id(:body), message_body_ui)
+      Yast::UI.ChangeWidget(Id(:back), :Enabled, !first_message?)
+      Yast::UI.ChangeWidget(Id(:next), :Enabled, !last_message?)
+    end
+
+    # Returns the current message
+    #
+    # That's the message that it's supposed to be shown to the user at a given
+    # time.
+    #
+    # @return Object
+    def current_message
+      messages[@position]
+    end
+
+    # Returns the amount of messages
+    #
+    # It is just a convenience method.
+    #
+    # @return [Integer] Amount of messages.
+    def total
+      messages.size
+    end
+
+    # UI content describing the UI for the current message's body
+    #
+    # @return [Yast::Term] UI content for dialog
+    def message_body_ui
+      title =
+        if messages.size > 1
+          format(_("%s (%d out of %d)"), current_message.title, @position + 1, total)
+        else
+          current_message.title
+        end
+      VBox(Label(title), RichText(current_message.body))
+    end
+
+    # Determines whether the current is the first message
+    #
+    # @return [true,false] True when it's the first message; true otherwise.
+    def first_message?
+      @position.zero?
+    end
+
+    # Determines whether the current is the last message
+    #
+    # @return [true,false] True when it's the last message; false otherwise.
+    def last_message?
+      @position == (total - 1)
+    end
+
+    # Build the set of buttons to be shown in the dialog
+    #
+    # Includes a 'stop' button only if the dialog is timed.
+    #
+    # @return [Yast::Term] UI content for dialog
+    def buttons_ui
+      buttons = [PushButton(Id(:close), Yast::Label.CloseButton)]
+      if messages.size > 1
+        buttons += [
+          HStretch(),
+          PushButton(Id(:back), Opt(:disabled), Yast::Label.BackButton),
+          HStretch(),
+          PushButton(Id(:next), Opt(last_message? ? :disabled : :enabled), Yast::Label.NextButton)
+        ]
+      end
+      if timed?
+        buttons.insert(2, PushButton(Id(:stop), Yast::Label.StopButton), HStretch())
+      end
+      HBox(*buttons)
```

Comment:
Maybe a `ButtonBox` should be used here instead of the `HBox`...

BTW: A nice trick with creating the button list! :+1: